src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+
+                    // The delay timer precision isn't exact. We might have a situation
+                    // where in the previous loop iterations intervalStartTime=20.00,
+                    // nextIntervalStartTime=21.00, the timer was supposed to delay for 1s but
+                    // it exited early so we looped around and DateTime.Now=20.99.
+                    // Aligning up from DateTime.Now would give us 21.00 again so we also need to skip
+                    // forward one time interval
+                    DateTime minNextInterval = intervalStartTime.AddSeconds(collectionIntervalSecs);
+                    if (nextIntervalStartTime <= minNextInterval)
+                    {
+                        nextIntervalStartTime = minNextInterval;
+                    }
+
+                    // pause until the interval is complete
+                    TimeSpan delayTime = nextIntervalStartTime - now;
+                    cancelToken.WaitHandle.WaitOne(delayTime);
+
+                    // don't do collection if timer may not have run to completion
+                    if (cancelToken.IsCancellationRequested)
+                    {
+                        break;
+                    }
+
+                    // collect statistics for the completed interval
+                    _beginCollection(intervalStartTime);
+                    Collect();
+                    _endCollection(intervalStartTime);
+                    intervalStartTime = nextIntervalStartTime;
+                }
+            }
+            catch (Exception e)
+            {
+                _collectionError(e);
+            }
+        }
+
+        public void Dispose()
+        {
+            _cts.Cancel();
+            if (_collectThread != null)
+            {
+                _collectThread.Join();
+                _collectThread = null;
+            }
+            _listener.Dispose();
+        }
+
+        private void RemoveInstrumentState(Instrument instrument, InstrumentState state)
+        {
+            _instrumentStates.TryRemove(instrument, out InstrumentState? _state);
+        }
+
+        private InstrumentState? GetInstrumentState(Instrument instrument)
+        {
+            if (!_instrumentStates.TryGetValue(instrument, out InstrumentState? instrumentState))
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
+                    {
+                        if (filter(instrument))
+                        {
+                            instrumentState = BuildInstrumentState(instrument);
+                            break;
+                        }
+                    }
+                }
+                if (instrumentState != null)
+                {
+                    _instrumentStates.TryAdd(instrument, instrumentState);
+                    instrumentState = _instrumentStates[instrument];
+                }
```

Comment:
would moving this block after the line 

instrumentState = BuildInstrumentState(instrument);

be better? or it is possible state be created in other places?

Reply:
I think there is no difference in correctness/perf so happy to swap it to the one that you felt was easier to read.